### PR TITLE
Infinispan client annotation marshaller support

### PR DESCRIPTION
**Pull request by wburns**
This commit will satisfy https://github.com/quarkusio/quarkus/issues/800

Now users can just add simple annotations to their classes and an Initializer class to automatically generate Marshaller(s) and proto schemas that are then automatically registered in the Injected RemoteCacheManager.

### CODE REVIEW

File: bom/runtime/pom.xml [hunk 1]
```diff
@@ -117,6 +117,7 @@
         <assertj.version>3.12.2</assertj.version>
         <json-smart.version>2.3</json-smart.version>
         <infinispan.version>10.0.0.Beta5</infinispan.version>
+        <infinispan.protostream.version>4.3.0.Alpha9</infinispan.protostream.version>
```
Comment (anistor) [inline bom/runtime/pom.xml]: Alpha10 was released yesterday.

File: docs/src/main/asciidoc/infinispan-client-guide.adoc [hunk 2]
```diff
@@ -99,15 +103,93 @@ public class Book {
...
+    public String getTitle() {
+        return title;
+    }
+
+    @ProtoField(number = 2)
+    public String getDescription() {
+        return description;
+    }
+
+    @ProtoField(number = 3, defaultValue = "1970")
```
Comment (anistor) [inline docs/src/main/asciidoc/infinispan-client-guide.adoc]: -1 is probably a better default to indicate the lack of a specified year. aka null :)

File: docs/src/main/asciidoc/infinispan-client-guide.adoc [hunk 3]
```diff
@@ -99,15 +103,93 @@ public class Book {
...
+
+    @ProtoField(number = 4)
+    public Set<Author> getAuthors() {
+        return authors;
+    }
+----
+If your classes have only mutable fields, then the `ProtoFactory` annotation
+is not required, assuming your class has a no arg constructor.
+
+Then all that is required is a very simple ContextInitializer interface with an annotation
```
Comment (anistor) [inline docs/src/main/asciidoc/infinispan-client-guide.adoc]: I think you actually mean ```SerializationContextInitializer```

File: docs/src/main/asciidoc/infinispan-client-guide.adoc [hunk 4]
```diff
@@ -99,15 +103,93 @@ public class Book {
...
+@AutoProtoSchemaBuilder(includeClasses = { Book.class, Author.class }, schemaPackageName = "book_sample")
+interface BookContextInitializer extends SerializationContextInitializer {
+}
+----
+
+So in this case we will automatically generate the marshaller and schemas for the included classes and
+place them in the schema package automatically. The package does not have to be provided, but if you
+utilize querying, you must know the generated package.
+
+NOTE: The `schemaFileName` and `schemaFilePath` attributes must NOT be set on the `AutoProtoSchemaBuilder` annotation. Failure to do so will cause native runtime to fail.
```
Comment (anistor) [inline docs/src/main/asciidoc/infinispan-client-guide.adoc]: I'd rephrase a a bit to clarifiy 'do not do this in Quakus context', not in general. Those attributes do work and have purpose outside Quarkus.

Comment (wburns) [reply docs/src/main/asciidoc/infinispan-client-guide.adoc]: Well I was trying to qualify it being that it fails during native runtime. As in it is fine in JVM mode with Quarkus.

File: docs/src/main/asciidoc/infinispan-client-guide.adoc (hunk 3 above)
Comment (wburns) [reply docs/src/main/asciidoc/infinispan-client-guide.adoc]: Sure, I was just making the word a bit smaller, but I can fully qualify :)

File: docs/src/main/asciidoc/infinispan-client-guide.adoc (hunk 2 above)
Comment (wburns) [reply docs/src/main/asciidoc/infinispan-client-guide.adoc]: Sure

File: bom/runtime/pom.xml (hunk 1 above)
Comment (wburns) [reply bom/runtime/pom.xml]: I will update.

File: docs/src/main/asciidoc/infinispan-client-guide.adoc (hunk 3 above)
Comment (wburns) [reply docs/src/main/asciidoc/infinispan-client-guide.adoc]: Sure, I was just making the word a bit smaller, but I can fully qualify :)

**Comment by wburns** [comment]
@anistor fixed review comments.

File: bom/runtime/pom.xml (hunk 1 above)
Comment (wburns) [reply bom/runtime/pom.xml]: Actually I can't update this until we release a new build of ISPN. It has to match the version that ISPN defines. I have reverted it back to Alpha9 until then.

File: docs/src/main/asciidoc/infinispan-client-guide.adoc [hunk 5]
```diff
@@ -99,15 +103,93 @@ public class Book {
...
+Serialization of user types uses a library based on protobuf, called Protostream.
 
-NOTE: Annotation based proto stream marshalling is not yet supported in
-the Quarkus Infinispan client. This will be added soon, allowing you to only annotate your classes,
-skipping the following steps.
+=== Annotation based Serialization
 
-NOTE: This version has an issue currently requiring the Query DSL module as a dependency when running in
-native mode with proto buf based marshalling. This should be remedied in a future version.
+This can be done automatically by adding annotations to your user classes from the protostream library
```
Comment (gsmet) [inline docs/src/main/asciidoc/infinispan-client-guide.adoc]: This sentence is unclear: `annotations to your user classes from the protostream library` .

Looks like the user classes are from the Protostream library, whereas I think you meant the annotations?

File: docs/src/main/asciidoc/infinispan-client-guide.adoc [hunk 6]
```diff
@@ -99,15 +103,93 @@ public class Book {
...
+provided by Infinispan. In addition a single Initializer annoated interface is required
```
Comment (gsmet) [inline docs/src/main/asciidoc/infinispan-client-guide.adoc]:
```suggestion
provided by Infinispan. In addition a single Initializer annotated interface is required
```

File: docs/src/main/asciidoc/infinispan-client-guide.adoc [hunk 7]
```diff
@@ -99,15 +103,93 @@ public class Book {
...
+
+    @ProtoField(number = 4)
+    public Set<Author> getAuthors() {
+        return authors;
+    }
+----
+If your classes have only mutable fields, then the `ProtoFactory` annotation
+is not required, assuming your class has a no arg constructor.
+
+Then all that is required is a very simple SerializationContextInitializer interface with an annotation
```
Comment (gsmet) [inline docs/src/main/asciidoc/infinispan-client-guide.adoc]:
```suggestion
Then all that is required is a very simple `SerializationContextInitializer` interface with an annotation
```

File: docs/src/main/asciidoc/infinispan-client-guide.adoc [hunk 8]
```diff
@@ -99,15 +103,93 @@ public class Book {
...
+@AutoProtoSchemaBuilder(includeClasses = { Book.class, Author.class }, schemaPackageName = "book_sample")
+interface BookContextInitializer extends SerializationContextInitializer {
+}
+----
+
+So in this case we will automatically generate the marshaller and schemas for the included classes and
+place them in the schema package automatically. The package does not have to be provided, but if you
+utilize querying, you must know the generated package.
+
+NOTE: In Quarkus the `schemaFileName` and `schemaFilePath` attributes should NOT be set on the `AutoProtoSchemaBuilder` annotation, failure to do so will cause native runtime to error.
```
Comment (gsmet) [inline docs/src/main/asciidoc/infinispan-client-guide.adoc]:
```suggestion
NOTE: In Quarkus the `schemaFileName` and `schemaFilePath` attributes should NOT be set on the `AutoProtoSchemaBuilder` annotation, setting them will cause native runtime to error.
```
I think it would be clearer this way.

File: docs/src/main/asciidoc/infinispan-client-guide.adoc [hunk 9]
```diff
@@ -303,20 +364,12 @@ The Infinispan client supports both indexed and non indexed querying as long as
 properties of the proto schema.
 
 Query builds upon the proto definitions you can configure when setting up the `ProtoStreamMarshaller`.
-Make sure to configure this marshaller before attempting querying. Once this is completed
-you must add another optional depdency: `infinispan-query-dsl` to start using queries.
+Either method of Serialization above will automatically register the schema with the server at
+startup, meaning that you will automatically gain the ability to query objects store with
```
Comment (gsmet) [inline docs/src/main/asciidoc/infinispan-client-guide.adoc]:
```suggestion
startup, meaning that you will automatically gain the ability to query the objects store with
```
but I'm not sure what you mean. Is it `the objects store`, `the objects stored`?

File: docs/src/main/asciidoc/infinispan-client-guide.adoc [hunk 10]
```diff
@@ -303,20 +364,12 @@ The Infinispan client supports both indexed and non indexed querying as long as
...
 
-[source,xml]
-----
-<dependency>
-    <groupId>org.infinispan</groupId>
-    <artifactId>infinispan-query-dsl</artifactId>
-</dependency>
-----
-
-With this added you can use Infinispan querying just as you would normally, nothing special required.
```
Comment (gsmet) [inline docs/src/main/asciidoc/infinispan-client-guide.adoc]: Either you don't remove that or you also remove the `You can do this by adding the following dependency to your project.` sentence but there's something wrong here.

**Review by gsmet — CHANGES_REQUESTED** [review]
I added a couple of comments on the documentation.

File: docs/src/main/asciidoc/infinispan-client-guide.adoc (hunk 5 above)
Comment (wburns) [reply docs/src/main/asciidoc/infinispan-client-guide.adoc]: Sure, I can cleanup.

File: docs/src/main/asciidoc/infinispan-client-guide.adoc (hunk 6 above)
Comment (wburns) [reply docs/src/main/asciidoc/infinispan-client-guide.adoc]: :+1:

File: docs/src/main/asciidoc/infinispan-client-guide.adoc (hunk 7 above)
Comment (wburns) [reply docs/src/main/asciidoc/infinispan-client-guide.adoc]: :+1:

File: docs/src/main/asciidoc/infinispan-client-guide.adoc (hunk 8 above)
Comment (wburns) [reply docs/src/main/asciidoc/infinispan-client-guide.adoc]: Sure.

File: docs/src/main/asciidoc/infinispan-client-guide.adoc (hunk 9 above)
Comment (wburns) [reply docs/src/main/asciidoc/infinispan-client-guide.adoc]: typo: supposed to be stored

File: docs/src/main/asciidoc/infinispan-client-guide.adoc (hunk 10 above)
Comment (wburns) [reply docs/src/main/asciidoc/infinispan-client-guide.adoc]: Oops missed removing the other.

File: docs/src/main/asciidoc/infinispan-client-guide.adoc (hunk 8 above)
Comment (wburns) [reply docs/src/main/asciidoc/infinispan-client-guide.adoc]: I changed it to `setting either`

Comment (wburns) [reply docs/src/main/asciidoc/infinispan-client-guide.adoc]: I changed it to `setting either`

**Comment by wburns** [comment]
Updated documentation.

**Comment by anistor** [comment]
LGTM @wburns ! But I think github still says there's one review comment left unaddressed.

**Comment by wburns** [comment]
There were two comments that I couldn't make outdated:
# Alpha9 - I can't upgrade until ISPN upgrades
# I removed `You can do this by adding the following dependency to your project` but it was on a different line.

**Comment by wburns** [comment]
To clarify all the review comments have been handled.

**Comment by anistor** [comment]
Computer says Noooo :)  "Requested changes must be addressed to merge this pull request."

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
The Windows failure looks unrelated, I'll merge this.
